mentions for secret blobs

mmckegg opened the pull request:
@dominictarr's [secret blobs proposal]() adds a query string onto the end of a blob ID containing the unbox key. These are not able to be detected by the mentions checker and so never get added to the `msg.value.content.mentions` array.

This PR fixes it by instead checking to see if the ID starts with a `&` and then assumes it is a blob. I did it this way because it was the same way @dominictarr did it ssbc/patchcore#50

I'm a little worried that we are veering of the "spec" established by `ssb-ref` here. Should we just update `ssb-ref` instead to be able to handle blob unbox in the id? 

Thoughts? 

cc @mixmix @clehner @ahdinosaur et el

mmckegg committed:
mentions for private blobs

mmckegg commented:
@clehner has already started work on this on git-ssb:

https://git.scuttlebot.io/%258%2BwcXeBD0H3Xv%2FoHJfhJ1mU4iryhhrztgdQBP77IRqA%3D.sha256/commits

It's a much better implementation that split out the `unbox=` into `key`.

mmckegg committed:
use ssb-ref with secret blob support for detecting blob mentions

mmckegg commented:
I just updated this PR to bring it in line with what @clehner's PR does, except that it uses an [updated version of ssb-ref](https://github.com/ssbc/ssb-ref/pull/15) that know how to parse secret blob refs.

dominictarr commented:
btw, I have been working on the ability to share a key to a encrypted message. This makes private messages into a capability system - you can create a private message, then reveal it later, or show it to another person. This is how I plan to implement private groups - create a new private message with the group key, then add more people to it.

https://github.com/auditdrivencrypto/private-box/commit/9849289a83d52043603e8e0e4b29d0a60792b7d8

(a particular message is the root, because I want it to always be explicit that you are adding someone to a group. it should not be possible to create a group that _just happens_ to have the same key as another group_)

This is related because it points towards needing a generic way to handle the passing a link with a decryption key.

ahdinosaur commented:
:+1: on `{name: name, link: link, query: { unbox: ... }}`

mmckegg committed:
use updated ssbc/ssb-ref#15 that handles any query on blobs

now matches @dominictarr's proposal on https://github.com/ssbc/ssb-mentions/pull/11#issuecomment-381442751

mmckegg committed:
parse out queries on all ref types (feed, msg, blob)

mmckegg commented:
Latest push adds `query` parsing to all ID mentions. You can now have a `%msgId?unbox=somecode&more=things` and it will get added as:

```
mentions: [{ 
  link: '%msgId', 
  query: {
    unbox: 'somecode', 
    more: 'things'
  }
}]
```

Shall I merge?

cc @dominictarr

mmckegg committed:
remove fork dep now that ssbc/ssb-ref#16 has merged

mmckegg commented:
Okay, I'm gonna merge!!
